internal/pkg/deploy/cloudformation/app.go

```diff
@@ -50,7 +52,66 @@ func (cf CloudFormation) DeployApp(in *deploy.CreateAppInput) error {
 }
 
 func (cf CloudFormation) UpgradeApplication(in *deploy.CreateAppInput) error {
-	return nil
+	appConfig := stack.NewAppStackConfig(in)
+	s, err := toStack(appConfig)
+	if err != nil {
+		return err
+	}
+	if err := cf.upgradeAppStack(s); err != nil {
+		return err
+	}
+	return cf.upgradeAppStackSet(appConfig)
+}
+
+func (cf CloudFormation) upgradeAppStackSet(config *stack.AppStackConfig) error {
+	for {
+		ssName := config.StackSetName()
+		if err := cf.appStackSet.WaitForStackSetLastOperationComplete(ssName); err != nil {
+			return fmt.Errorf("wait for stack set %s last operation complete: %w", ssName, err)
+		}
+		previouslyDeployedConfig, err := cf.getLastDeployedAppConfig(config)
+		if err != nil {
+			return err
+		}
+		previouslyDeployedConfig.Version += 1
+		err = cf.deployAppConfig(config, previouslyDeployedConfig)
+		if err == nil {
+			return nil
+		}
+		var stackSetOutOfDateErr *stackset.ErrStackSetOutOfDate
+		if errors.As(err, &stackSetOutOfDateErr) {
+			continue
+		}
+		return err
+	}
+}
+
+func (cf CloudFormation) upgradeAppStack(s *cloudformation.Stack) error {
+	for {
+		// Upgrade app stack.
+		descr, err := cf.cfnClient.Describe(s.Name)
+		if err != nil {
+			return fmt.Errorf("describe stack %s: %w", s.Name, err)
+		}
+		if cloudformation.StackStatus(aws.StringValue(descr.StackStatus)).InProgress() {
+			// There is already an update happening to the environment stack.
```

Comment:
was it intended to be "...happening to the application stack"?

Reply:
ooh sorry about it. you are right